### ISSUE TITLE
Handling JS alert/confirm/prompt

### ISSUE THREAD

**Issue by padde** [keathley/wallaby#169]
For example, when generating a scaffold with Phoenix, we get a confirmation dialog via `data-confirm` for "Delete" links in the index view. If this link is clicked, Wallaby/PhantomJS will *not* perform the actual request, then usually the next assertion or selector fails.

For now, I use [this workaround](https://github.com/detro/ghostdriver/issues/20#issuecomment-13621554), which I put before the part that triggers the alert/confirm:

```elixir
# for alert()
execute_script(session, "window.alert = function(){}")

# for confirm()
execute_script(session, "window.confirm = function(){return true;}")
```

However, it seems clunky to use the above workaround, even if extracted into a helper to reduce duplication. So, my questions are:

* Is there a way to ignore/auto-accept dialog boxes globally?
* Any better way of handling this that I didn't think of?
* Is there a way to interact with PhantomJS directly, so that we can implement proper dialog handling before https://github.com/detro/ghostdriver/issues/20 is fixed?  (i.e. supply our own `onAlert` etc. handlers directly to Phantom)
* Should we add documentation on how to handle these situations?

If I can do anything to help with implementation or documentation, I'd be glad to.

**Comment by padde** [comment]
So, after digging around a bit more I found that probably there is no nice way to handle this at the moment. The clean way would be to use the WebDriver interface to handle dialogs, but it seems that this won't happen soon.

The WebDriver spec [assumes that dialogs are blocking](https://www.w3.org/TR/webdriver/#user-prompts), so when a dialog appears the following command is expected to handle the dialog. If the subsequent command doesn't handle the dialog, an error is raised.

PhantomJS on the other hand assumes that dialogs are non-blocking. It expects a callback function such as [`page.onConfirm`](http://phantomjs.org/api/webpage/handler/on-confirm.html) before the actual dialog appears, where we can simulate user interaction with the dialog. This cannot trivially be made to block until the next command is received (accept/dismiss), therefore unlikely to be implemented soon in Ghostdriver. 

In addition to that, the WebDriver spec now also allows to specify a "user prompt behaviour" to globally handle dialogs. The necessary changes in Ghostdriver have already been implemented, and will hopefully be available once PhantomJS 2.5 is released. While this is good news, the solution has its own problems. For example, it seems impossible to specify a dynamic return value for a JS `prompt(message, default)`. Instead, [the default value is always returned](https://github.com/jesg/ghostdriver/blob/38c39cadf786f5446ae5c2a939798b6849338f2e/src/session.js#L386-L389).

All of this leads me to the conclusion that if we want to be able to handle prompts today, we need to have a way to provide the `onAlert`, `onConfirm` and `onPrompt` callbacks to PhantomJS directly. Luckily, Ghostdriver implements an endpoint `session/:session_id/phantom/execute` ([source](https://github.com/detro/ghostdriver/blob/7cf96db846e336a1a6dd9da891399902985dd57e/src/request_handlers/session_request_handler.js#L903-L910)). This works similarly to `execute_script`, but the injected code runs in the PhantomJS context, where `this` is the current `page` object. This can then be leveraged to add PhantomJS handlers.

I implemented a proof of concept [here](https://github.com/padde/wallaby/commit/f4b7594717d6e81d081f8e6b79e13fba4f9d2dac). If you would be open to adding `execute_phantom_script` to Wallaby that would be awesome, and I could make a proper PR out of this. I could also work on some helpers to set up callbacks, but I am not sure how such driver specific helpers would fit in with the overall architecture. I can perfectly understand if you'd rather keep those in a separate package, which I'd be happy to build as well 😉.

Please let me know what you think!

**Comment by keathley** [comment]
Thanks for putting so much research into this @padde! This is something that has caused issues with folks before. It looks like we might be able to incorporate your proof of concept into the init process of the phantom servers. We could provide a higher level api for handling alerts from the `Browser` module. Then the driver specific implementations can all go in the phantom driver.

I'd love to check out a PR for this if its something you wanna work on. Lemme know if there is anything that I can do to help as well.

**Comment by padde** [comment]
Thanks for the feedback @keathley. The only thing that's holding me back right now is that I want to get the API right before digging in all the way. I want to make sure we don't build a high level API that is incompatible with the WebDriver spec, which would then make it hard to implement additional drivers later.

With PhantomJS the way to handle alerts would be:

```elixir
session
|> visit("page.html")
|> accept_upcoming_alert()
|> click(Query.link("Trigger alert"))
```

whereas WebDriver assumes something like this:

```elixir
session
|> visit("page.html")
|> click(Query.link("Trigger alert"))
|> accept_currently_visible_alert()
```

If we want to implement these helpers today, we need to stick with the PhantomJS way. I am not sure how well other drivers would cope with this API though. We'd basically have to re-implement callbacks for those other drivers in Wallaby, keeping track of the registered prompt handlers etc, then figuring a way to trigger them at the right time. This would leave us with both special code for PhantomJS, and special code for the drivers that adhere to the WebDriver spec - not very pretty.

**Comment by PragTob** [comment]
(not sure my understanding of this is as whole as it could be)

Couldn't we just make a high level API that encompasses it like this:

```elixir
accept_alert_triggered_by fn -> click(session, Query.link("Trigger alert"))
```

then the web driver implementation could do the acceptance of the alert before or after calling the function depending on what the driver wants :)

**Comment by padde** [comment]
@PragTob that's a really nice idea! It would probably be useful to pass down the parent as an argument, so that we can write for example

```elixir
session
|> visit("page.html")
|> accept_alert(fn(session) ->
  click(session, Query.link("Trigger alert"))
end)
```

Maybe we could also add helpers to accept/dismiss all dialogs in case we're not really interested in specific ones. These could easily be implemented with the standard "unhandled prompt behaviour" from WebDriver if the driver implements them (e.g. will be in PhantomJS 2.5):

```elixir
session
|> accept_dialogs()
|> visit("page.html")
|> click(Query.link("Trigger alert"))
```

In summary, I'd then suggest the following API for `Browser`:

```elixir
@doc """
Accept all JavaScript dialogs.
Returns the `session`.
"""
def accept_dialogs(session)

@doc """
Dismiss all JavaScript dialogs.
Returns the `session`.
"""
def dismiss_dialogs(session)

@doc """
Accept one JavaScript `alert()` dialog triggered by the action within `fun`.
Returns the prompted text originally passed to `alert()`.
"""
def accept_alert(session, fun)

@doc """
Accept one JavaScript `confirm()` dialog triggered by the action within `fun`.
Returns the prompted text originally passed to `confirm()`.
"""
def accept_confirm(session, fun)

@doc """
Dismiss one JavaScript `confirm()` dialog triggered by the action within `fun`.
Returns the prompted text originally passed to `confirm()`.
"""
def dismiss_confirm(session, fun)

@doc """
Accept one JavaScript `prompt()` dialog triggered by the action within `fun`.
Will use the default value of the `prompt()` if no `return_value` is specified.
Returns the prompted text originally passed to `prompt()`.
"""
def accept_prompt(session, return_value \\ nil, fun)

@doc """
Dismiss one JavaScript `confirm()` dialog triggered by the action within `fun`.
Returns the prompted text originally passed to `prompt()`.
"""
def dismiss_prompt(session, fun)
```

The internal `Driver` API could probably be a bit more generic:

```
@doc """
Handle all subsequent JavaScript dialogs using the specified return value.
Returns the `session`.
"""
def handle_dialogs(session, return_value)

@doc """
Handle one JavaScript `alert()` dialog triggered by the action within `fun`.
Returns the prompted text originally passed to `alert()`.
"""
def handle_alert(session, fun)

@doc """
Handle one JavaScript `confirm()` dialog triggered by the action within `fun`.
Returns the prompted text originally passed to `confirm()`.
"""
def handle_confirm(session, return_value, fun)

@doc """
Handle one JavaScript `prompt()` dialog triggered by the action within `fun`.
Will use the default value of the `prompt()` if `return_value` is `nil`.
Returns the prompted text originally passed to `prompt()`.
"""
def handle_prompt(session, return_value, fun)
```

**Comment by padde** [comment]
One alternative to the above suggestion would be to return the `session` struct back from all helpers to keep them chainable, then have separate functions to retrieve the prompted text, for example:

```elixir
@doc """
Accept one JavaScript `alert()` dialog triggered by the action within `fun`.
Returns the `session`.
"""
def accept_alert(session, fun)

# ...

@doc """
Returns all messages from JavaScript dialogs handled with `accept_alert` etc., which were triggered within the specified `session`.
"""
def dialog_messages(session, fun)

@doc """
Returns the last message from a JavaScript dialog handled with `accept_alert` etc., which was triggered within the specified `session`.
"""
def last_dialog_message(session, fun)
```

**Comment by keathley** [comment]
Closing since this has been merged with #173

**Comment by btbinhtran** [comment]
So I tried using the `accept_confirm` function to handle a confirm box after clicking a "delete" link on a user, but the item does not get deleted. I am using Wallaby 0.22 and `Wallaby.Experimental.Chrome` driver.

However if I use @padde original work around
```elixir
execute_script(session, "window.confirm = function(){return true;}")
```

before clicking the link, everything works as expected.

Here's a test I'm writing. (The commented code would make the last `assert` fail)
```elixir
defmodule SampleAppWeb.UserIndexTest do
  use SampleAppWeb.IntegrationCase, async: true
  import Wallaby.Query
  alias SampleApp.Accounts
  alias SampleApp.Accounts.User
  alias SampleAppWeb.Endpoint
  alias SampleApp.Factory

  setup do
    admin = Factory.insert(:user, admin: true)
    non_admin = Factory.insert(:user)
    for _n <- 1..35 do
      Factory.insert(:user)
    end
    {:ok, admin: admin,
          non_admin: non_admin}
  end

  test "index as admin including pagination and delete links",
    %{session: session, admin: admin, non_admin: non_admin} do
    session
    |> login_as(admin)
    |> visit(user_path(Endpoint, :index))
    |> assert_has(css("ul.pagination", count: 2))
    first_page_of_users = Accounts.paginate_users(%{page: 1})
    for user <- first_page_of_users do
      session
      |> assert_has(css("a[href='#{user_path(Endpoint, :show, user)}']", text: user.name))
      unless user.id == admin.id do
        session
        |> assert_has(css("a[href='#{user_path(Endpoint, :delete, user)}'][data-method='delete']"))
      end
    end
    user_records_before = Repo.one(from u in User, select: count())
    
    session
    |> execute_script("window.confirm = function(){return true;}")
    |> click(css("a[href='#{user_path(Endpoint, :delete, non_admin)}'][data-method='delete']"))
    
    # message = accept_confirm(session, fn(s) ->
    #   click(s, css("a[href='#{user_path(Endpoint, :delete, non_admin)}'][data-method='delete']"))
    # end)

    user_records_after = Repo.one(from u in User, select: count())
    assert user_records_before == user_records_after + 1
  end
end
```
